From 759c8a9e6e4223cb1f8a02480a8a7b1ffa1f05b5 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Tue, 23 Sep 2014 19:02:04 -0400
Subject: [fs] nfs: remove BUG possibility in nfs4_open_and_get_state

Message-id: <1411498926-30119-2-git-send-email-steved@redhat.com>
Patchwork-id: 96498
O-Subject: [RHEL7.1 PATCH 1/3] NFS: remove BUG possibility in nfs4_open_and_get_state
Bugzilla: 1111708
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: NeilBrown <neilb@suse.de>

commit 4fa2c54b5198d09607a534e2fd436581064587ed
    NFS: nfs4_do_open should add negative results to the dcache.

used "d_drop(); d_add();" to ensure that a dentry was hashed
as a negative cached entry.
This is not safe if the dentry has an non-NULL ->d_inode.
It will trigger a BUG_ON in d_instantiate().
In that case, d_delete() is needed.

Also, only d_add if the dentry is currently unhashed, it seems
pointless removed and re-adding it unchanged.

Reported-by: Christoph Hellwig <hch@infradead.org>
Fixes: 4fa2c54b5198d09607a534e2fd436581064587ed
Cc: Jeff Layton <jeff.layton@primarydata.com>
Link: http://lkml.kernel.org/r/20140908144525.GB19811@infradead.org
Signed-off-by: NeilBrown <neilb@suse.de>
Acked-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: f39c01047994e66e7f3d89ddb4c6141f23349d8d
Bugzilla: 1111708

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/nfs4proc.c b/fs/nfs/nfs4proc.c
index 2e9a0b5..21b4e5d 100644
--- a/fs/nfs/nfs4proc.c
+++ b/fs/nfs/nfs4proc.c
@@ -2226,9 +2226,13 @@ static int _nfs4_open_and_get_state(struct nfs4_opendata *opendata,
  ret = _nfs4_proc_open(opendata);
  if (ret != 0) {
   if (ret == -ENOENT) {
-   d_drop(opendata->dentry);
-   d_add(opendata->dentry, NULL);
-   nfs_set_verifier(opendata->dentry,
+   dentry = opendata->dentry;
+   if (dentry->d_inode)
+    d_delete(dentry);
+   else if (d_unhashed(dentry))
+    d_add(dentry, NULL);
+
+   nfs_set_verifier(dentry,
       nfs_save_change_attribute(opendata->dir->d_inode));
   }
   goto out;
-- 
1.7.1